ext/imap/php_imap.c

```diff
@@ -138,28 +140,44 @@ ZEND_TSRMLS_CACHE_DEFINE()
 ZEND_GET_MODULE(imap)
 #endif
 
-/* True globals, no need for thread safety */
-static int le_imap;
+/* Imap class entry definition */
+static zend_class_entry *php_imap_ce = NULL;
+static zend_object_handlers imap_object_handlers;
 
-// TODO Promote to ValueError?
-#define PHP_IMAP_CHECK_MSGNO(msgindex, arg_pos)	\
-	if (msgindex < 1) { \
-		zend_argument_value_error(arg_pos, "must be greater than 0"); \
-		RETURN_THROWS(); \
-	} \
-	if (((unsigned) msgindex) > imap_le_struct->imap_stream->nmsgs) { \
-		php_error_docref(NULL, E_WARNING, "Bad message number"); \
-		RETURN_FALSE;	\
-	}	\
+typedef struct _php_imap_object {
+	MAILSTREAM *imap_stream;
+	long flags;
+	zend_object std;
+} php_imap_object;
 
-/* {{{ mail_close_it */
-static void mail_close_it(zend_resource *rsrc)
-{
-	pils *imap_le_struct = (pils *)rsrc->ptr;
+static inline zend_object *imap_object_to_zend_object(php_imap_object *obj) {
+	return ((zend_object*)(obj + 1)) - 1;
+}
 
-	/* Do not try to close prototype streams */
-	if (!(imap_le_struct->flags & OP_PROTOTYPE)) {
-		mail_close_full(imap_le_struct->imap_stream, imap_le_struct->flags);
+static inline php_imap_object *imap_object_from_zend_object(zend_object *zobj) {
+	return ((php_imap_object*)(zobj + 1)) - 1;
```

Comment:
This has basically the same issue as above, but to convert from `zend_object*` to `php_imap_object*` one usually [uses the `XtOffsetOf` macro](https://github.com/php/php-src/blob/master/ext/xml/xml.c#L345).

Reply:
And again, because they're not the same size structs, the same answer applies.

XtOffsetOf() does the same math presented in my boilerplate helpers, but doesn't present the (honestly kinda ugly) hack of intermediately casting to char*.  c'mon, that's gross.

Reply:
But that's how [it's documented](https://wiki.php.net/phpng-upgrading?s[]=xtoffsetof#custom_objects). :)

Reply:
Eh. Lots of documentation is wrong about lots of things. ;)